In [6]:
from pedalboard import Pedalboard, Reverb, load_plugin
from pedalboard.io import AudioFile
from mido import Message # not part of Pedalboard, but convenient!
# import timeit
import psutil
import time
import cProfile
# from memory_profiler import profile
import tracemalloc


In [7]:
# from ctypes import cdll
# lib_sde_markers = cdll.LoadLibrary('./libsde_markers.so')

In [8]:
# !pip install memory_profiler
# !pip install psutil
# !pip install flops-profiler

In [9]:
# def count_flops(func):
#     def wrapper(*args, **kwargs):
#         # Assuming each operation counts as 1 FLOP
#         flops = 0
#         # Example operation counting (customize based on your function)
#         flops += 1000  # hypothetical count for demonstration
        
#         start_time = time.perf_counter()
#         result = func(*args, **kwargs)
#         end_time = time.perf_counter()
        
#         elapsed_time = end_time - start_time
#         tflops = flops / (elapsed_time * 1e12)
        
#         print(f"FLOPs: {flops}, Execution Time: {elapsed_time:.6f}s, TFLOPS: {tflops:.6f}")
#         return result
#     return wrapper

In [12]:
# Load a VST3 or Audio Unit plugin from a known path on disk:
# instrument = load_plugin("./VSTs/Magical8BitPlug2.vst3")
def apply_uvi_tape_flanger_effect(src_wav=None, output_wav=None):
    effect = load_plugin("../../../sound_effects/UVI Tape Flanger.vst3")
    # for e in effect.parameters.items():
    #     print(e)
    effect.ration = 15;
    effect.apply_tape_simulation = True
    effect.delay = '12.0 ms';
    with AudioFile(src_wav) as f:
  # Open an audio file to write to:
      with AudioFile(output_wav, 'w', f.samplerate, f.num_channels) as o:
         # Read one second of audio at a time, until the file is empty:
        while f.tell() < f.frames:
          chunk = f.read(f.samplerate)
          # Run the audio through our pedalboard:
          effected = effect(chunk, f.samplerate, reset=False)
          # Write the output to our output file:
          o.write(effected)

In [13]:
apply_uvi_tape_flanger_effect("test02.wav","./output_wav/output_uvi_flanger.wav")

In [11]:
# apply_uvi_tape_flanger_effect('test02.wav','./output_wav/output_uvi_flanger.wav')
cProfile.run('apply_uvi_tape_flanger_effect("test02.wav","./output_wav/output_uvi_flanger.wav")')

         2472344 function calls (2472284 primitive calls) in 3.739 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    1.635    1.635    3.739    3.739 462557544.py:3(apply_uvi_tape_flanger_effect)
        1    0.000    0.000    3.739    3.739 <string>:1(<module>)
    56056    0.052    0.000    0.412    0.000 _pedalboard.py:129(looks_like_float)
    92092    0.450    0.000    0.450    0.000 _pedalboard.py:242(get_text_for_raw_value)
       92    0.235    0.003    2.023    0.022 _pedalboard.py:284(__init__)
       52    0.000    0.000    0.000    0.000 _pedalboard.py:307(<genexpr>)
    56122    0.026    0.000    0.437    0.000 _pedalboard.py:326(<genexpr>)
       58    0.057    0.001    0.412    0.007 _pedalboard.py:328(<dictcomp>)
       58    0.028    0.000    0.065    0.001 _pedalboard.py:360(<listcomp>)
       26    0.000    0.000    0.000    0.000 _pedalboard.py:370(<setcomp>)
       24    0.000    0.000    0.0

In [9]:
# %timeit apply_uvi_tape_flanger_effect('test02.wav','./output_wav/output_uvi_flanger.wav')
# start_time = time.perf_counter()  # Start time measurement
## measure memory needed for python function
# tracemalloc.start()  # Start tracking memory allocations
# apply_uvi_tape_flanger_effect('test02.wav','./output_wav/output_uvi_flanger.wav')
# snapshot = tracemalloc.take_snapshot()
# top_stats = snapshot.statistics('lineno')  # Get statistics by line number
# for stat in top_stats[:10]:  # Print top 10 lines consuming memory
#     print(stat)
# tracemalloc.stop()  # Stop tracking

# end_time = time.perf_counter()  # End time measurement
# elapsed_time = end_time - start_time  # Calculate elapsed time in seconds
# print(f"Elapsed time: {elapsed_time:.6f} seconds")

### cProfile
2472344 function calls (2472284 primitive calls) in 3.739 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    1.635    1.635    3.739    3.739 462557544.py:3(apply_uvi_tape_flanger_effect)
        1    0.000    0.000    3.739    3.739 <string>:1(<module>)
    56056    0.052    0.000    0.412    0.000 _pedalboard.py:129(looks_like_float)
    92092    0.450    0.000    0.450    0.000 _pedalboard.py:242(get_text_for_raw_value)
       92    0.235    0.003    2.023    0.022 _pedalboard.py:284(__init__)
       52    0.000    0.000    0.000    0.000 _pedalboard.py:307(<genexpr>)
    56122    0.026    0.000    0.437    0.000 _pedalboard.py:326(<genexpr>)
       58    0.057    0.001    0.412    0.007 _pedalboard.py:328(<dictcomp>)
       58    0.028    0.000    0.065    0.001 _pedalboard.py:360(<listcomp>)
       26    0.000    0.000    0.000    0.000 _pedalboard.py:370(<setcomp>)
       24    0.000    0.000    0.000    0.000 _pedalboard.py:373(<dictcomp>)
       92    0.011    0.000    0.011    0.000 _pedalboard.py:383(<dictcomp>)
     2966    0.017    0.000    0.017    0.000 _pedalboard.py:385(__get_cpp_parameter)
       58    0.000    0.000    0.002    0.000 _pedalboard.py:402(label)
       59    0.000    0.000    0.001    0.000 _pedalboard.py:468(__getattr__)
     1666    0.008    0.000    0.033    0.000 _pedalboard.py:479(__setattr__)
        2    0.000    0.000    0.000    0.000 _pedalboard.py:489(get_raw_value_for)
       92    0.001    0.000    0.004    0.000 _pedalboard.py:592(to_python_parameter_name)
       92    0.001    0.000    0.003    0.000 _pedalboard.py:602(normalize_python_parameter_name)
       92    0.001    0.000    0.002    0.000 _pedalboard.py:607(<listcomp>)
       92    0.000    0.000    0.000    0.000 _pedalboard.py:614(<listcomp>)
        1    0.000    0.000    1.020    1.020 _pedalboard.py:623(__set_initial_parameter_values__)
        1    0.000    0.000    1.020    1.020 _pedalboard.py:644(parameters)
        2    0.002    0.001    2.026    1.013 _pedalboard.py:650(_get_parameters)
       92    0.001    0.000    0.001    0.000 _pedalboard.py:658(<listcomp>)
        3    0.000    0.000    1.006    0.335 _pedalboard.py:670(_get_parameter_by_python_name)
        4    0.000    0.000    0.000    0.000 _pedalboard.py:698(__getattr__)
      7/5    0.000    0.000    1.006    0.201 _pedalboard.py:721(__setattr__)
        1    0.077    0.077    1.098    1.098 _pedalboard.py:753(load_plugin)
   164230    0.782    0.000    1.014    0.000 _pedalboard.py:97(strip_common_float_suffixes)
     1483    0.003    0.000    0.003    0.000 contextlib.py:102(__init__)
     1483    0.001    0.000    0.019    0.000 contextlib.py:130(__enter__)
     1483    0.002    0.000    0.003    0.000 contextlib.py:139(__exit__)
     1483    0.001    0.000    0.004    0.000 contextlib.py:279(helper)
   220198    0.031    0.000    0.031    0.000 typing.py:1737(cast)
    54000    0.008    0.000    0.008    0.000 {built-in method builtins.abs}
      118    0.010    0.000    0.447    0.004 {built-in method builtins.all}
       92    0.000    0.000    0.000    0.000 {built-in method builtins.any}
        1    0.000    0.000    3.739    3.739 {built-in method builtins.exec}
     1488    0.000    0.000    0.000    0.000 {built-in method builtins.getattr}
165688/165630    0.023    0.000    0.023    0.000 {built-in method builtins.hasattr}
   328621    0.046    0.000    0.046    0.000 {built-in method builtins.isinstance}
      160    0.000    0.000    0.000    0.000 {built-in method builtins.iter}
   492826    0.055    0.000    0.055    0.000 {built-in method builtins.len}
       58    0.001    0.000    0.001    0.000 {built-in method builtins.max}
       58    0.001    0.000    0.001    0.000 {built-in method builtins.min}
     3126    0.001    0.000    0.018    0.000 {built-in method builtins.next}
     1080    0.000    0.000    0.000    0.000 {built-in method builtins.ord}
    54000    0.030    0.000    0.030    0.000 {built-in method builtins.round}
        2    0.000    0.000    0.000    0.000 {built-in method builtins.setattr}
       58    0.001    0.000    0.001    0.000 {built-in method builtins.sorted}
       14    0.000    0.000    0.000    0.000 {built-in method builtins.sum}
    54058    0.010    0.000    0.010    0.000 {method 'add' of 'set' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
   110114    0.034    0.000    0.034    0.000 {method 'endswith' of 'str' objects}
       58    0.000    0.000    0.000    0.000 {method 'format' of 'str' objects}
        4    0.000    0.000    0.000    0.000 {method 'get' of 'dict' objects}
     1172    0.000    0.000    0.000    0.000 {method 'isalpha' of 'str' objects}
       92    0.000    0.000    0.000    0.000 {method 'isnumeric' of 'str' objects}
     1080    0.000    0.000    0.000    0.000 {method 'isprintable' of 'str' objects}
      175    0.000    0.000    0.000    0.000 {method 'items' of 'dict' objects}
       92    0.000    0.000    0.000    0.000 {method 'join' of 'str' objects}
   164422    0.024    0.000    0.024    0.000 {method 'lower' of 'str' objects}
      184    0.000    0.000    0.000    0.000 {method 'match' of 're.Pattern' objects}
    54334    0.017    0.000    0.017    0.000 {method 'replace' of 'str' objects}
     1736    0.001    0.000    0.001    0.000 {method 'startswith' of 'str' objects}
   382642    0.053    0.000    0.053    0.000 {method 'strip' of 'str' objects}
      526    0.000    0.000    0.000    0.000 {method 'values' of 'dict' objects}

### uvi tape flanger effect
- Elapsed time: 2.755284 seconds
- Memory needed:
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:303: size=3300 KiB, count=57711, average=59 B
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:383: size=2140 KiB, count=145, average=14.8 KiB
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:368: size=1994 KiB, count=116, average=17.2 KiB
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:297: size=1359 KiB, count=57970, average=24 B
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:329: size=1267 KiB, count=54058, average=24 B
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:381: size=459 KiB, count=184, average=2555 B
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:256: size=323 KiB, count=3962, average=84 B
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:487: size=56.7 KiB, count=175, average=332 B
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:661: size=14.8 KiB, count=187, average=81 B
  - /Users/xrickliao/miniconda3/envs/dlenv1/lib/python3.10/site-packages/pedalboard/_pedalboard.py:667: size=10.0 KiB, count=95, average=108 B

- total memory needed for apply flanger effect: 3300+2140+1994+1359+1267+459+323+56.7+14.8+10 = 8929.5KB ~ 9MB


In [5]:
# # Open an audio file for reading, just like a regular file:
# with AudioFile('test02.wav') as f:
  
#   # Open an audio file to write to:
#   with AudioFile('output.wav', 'w', f.samplerate, f.num_channels) as o:
  
#     # Read one second of audio at a time, until the file is empty:
#     while f.tell() < f.frames:
#       chunk = f.read(f.samplerate)
#       # Run the audio through our pedalboard:
#       effected = board(chunk, f.samplerate, reset=False)
#       # Write the output to our output file:
#       o.write(effected)